docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,96 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original node key of IAVL is a hash of the node and it does not take advantage of data locality on disk. The nodes are stored in a random location of the disk due to the random hash value, so it needs to do a random search of the disk to find the node.
+
+The `orphans` are used to manage the removed nodes in the current version and allow to deletion of the removed nodes for the specific version from the disk through the `DeleteVersion`. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases, the first one is the rollback of the tree and the second one is to remove the unnecessary old nodes.
+
+## Decision
+
+- Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
+- Remove the `version` field from the node structure.
+- Remove the `orphans` from the tree.
+
+New node structure
+
+```go
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte    // keep this field in the storage
+	leftHash      []byte    // will remove
+	rightHash     []byte    // will remove
+	nodeKey       int64     // new field, use as a node key
+	leftNodeKey   int64     // new field, need to store in the storage
+	rightNodeKey  int64     // new field, need to store in the storage
+	version       int64     // will remove
+	size          int64
+	leftNode      *Node
+	rightNode     *Node
+	subtreeHeight int8
+	persisted     bool
+}
+```
+
+New tree structure
+
+```go
+type MutableTree struct {
+	*ImmutableTree                                    
+	lastSaved                *ImmutableTree
+	nonce                    int64                    // new field to track the current ID
+	orphans                  map[int64]int64          // will remove
+	versions                 map[int64]bool           
+	allRootLoaded            bool                     
+	unsavedFastNodeAdditions map[string]*fastnode.Node
+	unsavedFastNodeRemovals  map[string]interface{}   
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool 
+
+	mtx sync.Mutex
+}
+```
+
+## Consequences
+
+### Migration
+
+We can migrate nodes one by one by iterating the version.
+
+- Iterate the version in order, and get the root node for the specific version.
```

Comment:
this may need a larger discussion but should it be lazily or stop the node migrate then restart? Any thoughts on that?

Reply:
Re-assign should be done after stopping the node. Maybe we can delete only nodes which not appear in iterating. It can be done without breaking off. I will introduce both ways.

Reply:
I can grab some snapshots to test this migration and the time it would take. It's gonna be a long one.

Reply:
Sorry, I was confused with pruning!
I think lazy loading is possible but we should restart the node to switch database after loading, right?

Reply:
hmm, maybe lazy migration is not possible. cause it needs to start from the lowest version in the db then move upwards, right?

Reply:
I think it is possible, but this should be done at the top level, not within iavl.
For example, we load it to the new IAVL lazily through `Export` of the original version and the new `Import` functionality, after syncing and block confirmation, we can switch the store to the new version.

Reply:
yes that also works. not sure how that would work with go.mods though unless we call this v1 so we can import both?

Reply:
https://github.com/golang/go/wiki/Modules#version-selection
Maybe, we should create v2 of iavl.